### PR TITLE
Add bond support on cisco

### PR DESCRIPTION
**Pull request by JoProvost**
- Add bond
 - Remove bond
 - Get bond
 - Get bonds
 - Add interface to bond
 - Remove interface from bond

### CODE REVIEW

File: tests/adapters/switches/cisco_test.py [hunk 1]
```diff
@@ -2429,3 +2427,311 @@ def test_remove_non_existent_dhcp_relay_server_fails(self):
...
+            "",
+            "Number of channel-groups in use: 1",
+            "Number of aggregators:           1",
+            "",
+            "Group  Port-channel  Protocol    Ports",
+            "------+-------------+-----------+-----------------------------------------------",
+            "1      Po1(SU)         LACP      Fa0/1(P)",
+            "2      Po2(SU)         LACP      Gi1/1(P) Gi1/2(P)",
+            "3      Po3(S)          LACP      ",
+            "",
```
Comment (lindycoder) [inline tests/adapters/switches/cisco_test.py]: This is an error, ssh_clients are made to not return empty lines

File: netman/adapters/switches/cisco.py [hunk 2]
```diff
@@ -137,10 +139,60 @@ def get_interfaces(self):
...
             interface = parse_interface(data)
-            if interface:
+            if interface is not None and isinstance(interface, Interface):
                 interfaces.append(interface)
         return interfaces
 
+    def get_bonds(self):
+        bonds = []
+        for bond_summary in self._show_etherchannel_summary():
+            if bond_summary:
```
Comment (lindycoder) [inline netman/adapters/switches/cisco.py]: you don't need this if, empty lines are not returned by the shell clients

File: netman/adapters/switches/cisco.py [hunk 3]
```diff
@@ -137,10 +139,60 @@ def get_interfaces(self):
...
+                bonds.append(self._bond_from_summary(bond_summary))
+        return bonds
+
+    def get_bond(self, number):
+        for bond_summary in self._show_etherchannel_summary():
+            if bond_summary and int(bond_summary[0]) == number:
```
Comment (lindycoder) [inline netman/adapters/switches/cisco.py]: same here

File: netman/adapters/switches/cisco.py [hunk 4]
```diff
@@ -398,34 +450,53 @@ def remove_vrrp_group(self, vlan_number, group_id):
...
+        return split_lines_after('---', self.ssh.do("show etherchannel summary"))
+
+    def _bond_from_summary(self, bond_summary):
+        bond = self.get_interface(bond_summary[1].rsplit('(')[0])
+        bond.members = [self.get_interface(m.rsplit('(')[0]).name for m in bond_summary[3:]]
+        return bond
+
+    def _has_bond(self, number):
+        for bond_summary in self._show_etherchannel_summary():
+            if bond_summary and int(bond_summary[0]) == number:
```
Comment (lindycoder) [inline netman/adapters/switches/cisco.py]: same, first part of the if is not needed

File: netman/adapters/switches/cisco.py [hunk 5]
```diff
@@ -137,10 +139,60 @@ def get_interfaces(self):
...
+                return self._bond_from_summary(bond_summary)
+        raise UnknownBond(number)
+
+    def add_bond(self, number):
+        if self._has_bond(number):
+            raise BondAlreadyExist(number)
+        with self.config():
+            result = self.ssh.do('interface port-channel {}'.format(number))
```
Comment (lindycoder) [inline netman/adapters/switches/cisco.py]: you could starting unsing the result checker introduced in the dell here

https://github.com/internaphosting/netman/blob/master/netman/adapters/switches/dell.py#L258

just a proposition

File: netman/adapters/switches/cisco.py [hunk 6]
```diff
@@ -137,10 +139,60 @@ def get_interfaces(self):
...
+    def remove_bond(self, number):
+        if not self._has_bond(number):
+            raise UnknownBond(number)
+        with self.config():
+            self.ssh.do('no interface port-channel {}'.format(number))
+
+    def add_interface_to_bond(self, interface, bond_number):
+        if not self._has_bond(bond_number):
+            raise UnknownBond(bond_number)
+        with self.config():
```
Comment (lindycoder) [inline netman/adapters/switches/cisco.py]: why not in a self.interface() instead of manually exiting at the end?

File: netman/adapters/switches/cisco.py [hunk 7]
```diff
@@ -137,10 +139,60 @@ def get_interfaces(self):
...
+            if len(result) > 0:
+                raise UnknownInterface(interface)
+            self.ssh.do('channel-protocol lacp')
+            self.ssh.do('channel-group {} mode active'.format(bond_number))
+            if len(result) > 0:
+                raise BadBondNumber()
+            self.ssh.do('exit')
+
+    def remove_interface_from_bond(self, interface):
+        with self.config():
```
Comment (lindycoder) [inline netman/adapters/switches/cisco.py]: here too, using self.interface would launch the exception for you no?

File: tests/adapters/compliance_tests/add_bond_test.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+        bonds = self.client.get_bonds()
+        assert_that(bonds, has_item(has_properties(number=10)))
+
+        bond = next((b for b in bonds if b.number == 10))
+        assert_that(bond.number, is_(10))
+        assert_that(bond.members, is_([]))
+        assert_that(bond.link_speed, is_(None))
+        assert_that(bond.shutdown, is_(False))
+        assert_that(bond.port_mode, is_(DYNAMIC))
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/add_bond_test.py]: This should be access, i think the DYNAMIC port mode should be removed and cisco ports forced into access, if possible for unified behavior caus i don't think other switches we support have such a mode

File: tests/adapters/compliance_tests/__init__.py [hunk 9]
```diff
@@ -23,7 +24,8 @@
 _test_classes = [
     GetVlanTest,
     AddVlanTest,
-    GetInterfacesTest
+    GetInterfacesTest,
+    AddBondTest,
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/__init__.py]: AddInterfaceToBondTest is missing here

Comment (lindycoder) [inline tests/adapters/compliance_tests/__init__.py]: RemoveInterfaceFromBondTest too

File: tests/adapters/compliance_tests/add_interface_to_bond_test.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+from hamcrest import assert_that, is_, none, empty, has_item, has_properties
+from netman.core.objects.exceptions import UnknownBond
+
+from tests.adapters.configured_test_case import ConfiguredTestCase
+
+
+class AddInterfaceToBondTest(ConfiguredTestCase):
+    _dev_sample = "cisco"
+
+    def test_adding_an_interface_to_a_bond(self):
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/add_interface_to_bond_test.py]: The sentence : 
AddInterfaceToBondTest adding an interface to a bond
is not readable, sorry to be an ass about that
Proposition :
test_adds_the_interface_to_the_bond

it's a pedantic issue, yes but if we can manage to have an auto-documentation with these it would be awesome :)

File: tests/adapters/compliance_tests/add_interface_to_bond_test.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        self.client.add_bond(9)
+
+        # when
+        self.client.add_interface_to_bond(bond_member, 9)
+
+        # then
+        bond = self.client.get_bond(9)
+        assert_that(bond.members, is_([]))
+
+        interfaces = self.client.get_interfaces()
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/add_interface_to_bond_test.py]: may i suggest a method along the line of 

https://github.com/internaphosting/netman/blob/master/tests/adapters/configured_test_case.py#L95

File: tests/adapters/compliance_tests/add_interface_to_bond_test.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+class AddInterfaceToBondTest(ConfiguredTestCase):
+    _dev_sample = "cisco"
+
+    def test_adding_an_interface_to_a_bond(self):
+        # given
+        bond_member = self.test_ports[0].name
+        self.client.add_bond(9)
+
+        # when
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/add_interface_to_bond_test.py]: the given when and then are not big enough benefit from these seperation IMHO, i don't think they really help understand the flow of the test, juste a line break is enough i think

File: tests/adapters/compliance_tests/add_bond_test.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        assert_that(bond.access_vlan, is_(None))
+        assert_that(bond.trunk_native_vlan, is_(None))
+        assert_that(bond.trunk_vlans, is_([]))
+
+    def test_fails_if_the_bond_already_exist(self):
+        self.client.add_bond(10)
+
+        with self.assertRaises(BondAlreadyExist):
+            self.client.add_bond(10)
+
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/add_bond_test.py]: test_fails_with_an_invalid_number  could be relevant

File: tests/adapters/compliance_tests/add_interface_to_bond_test.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        interfaces = self.client.get_interfaces()
+        assert_that(interfaces, has_item(has_properties(name=bond_member)))
+        interface = next((b for b in interfaces if b.name == bond_member))
+        assert_that(interface.bond_master, is_(9))
+
+    def test_fails_if_the_bond_does_not_exist(self):
+        bond_member = self.test_ports[0].name
+        with self.assertRaises(UnknownBond):
+            self.client.add_interface_to_bond(bond_member, 9)
+
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/add_interface_to_bond_test.py]: fails if the interface does not exist

Comment (lindycoder) [inline tests/adapters/compliance_tests/add_interface_to_bond_test.py]: What should be the behavior if the interface is already inside the bond?   Fail or not?

File: tests/adapters/compliance_tests/remove_interface_from_bond_test.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from hamcrest import assert_that, is_, none, empty, has_item, has_properties
+from netman.core.objects.exceptions import UnknownBond
+
+from tests.adapters.configured_test_case import ConfiguredTestCase
+
+
+class RemoveInterfaceFromBondTest(ConfiguredTestCase):
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/remove_interface_from_bond_test.py]: Most of the comment for the add interface to bond should be addressed in this class too

File: tests/adapters/compliance_tests/remove_interface_from_bond_test.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+        # then
+        bond = self.client.get_bond(9)
+        assert_that(bond.members, is_([]))
+
+        interfaces = self.client.get_interfaces()
+        assert_that(interfaces, has_item(has_properties(name=bond_member)))
+        interface = next((b for b in interfaces if b.name == bond_member))
+        assert_that(interface.bond_master, is_(None))
+
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/remove_interface_from_bond_test.py]: What is the behavior if the interface is not already inside the bond when removing? fail or not?

It is a resource in a list, therefore i think it should fail

Comment (lindycoder) [inline tests/adapters/compliance_tests/remove_interface_from_bond_test.py]: invalid interface should also be covered

File: tests/adapters/switches/cisco_test.py [hunk 17]
```diff
@@ -53,12 +53,12 @@ def setUp(self):
         self.lock = mock.Mock()
         self.switch = cisco.factory(SwitchDescriptor(model='cisco', hostname="my.hostname", password="the_password"), self.lock)
         SubShell.debug = True
+        self.mocked_ssh_client = flexmock()
```
Comment (lindycoder) [inline tests/adapters/switches/cisco_test.py]: Why this change? all the new tests use command_setup

**Comment by idjaw** [comment]
What should be done with this pull request? Is this work that should be eventually merged, or should we close this pr and get back to it when ready to revisit?

**Comment by fbouliane** [comment]
:+1: for close & reopen when rebased and fixed.

**Comment by JoProvost** [comment]
Will work on this later, when required
